apps/dc_tools/odc/apps/dc_tools/utils.py

```diff
@@ -156,52 +166,100 @@ def index_update_dataset(
     update=False,
     update_if_exists=False,
     allow_unsafe=False,
-):
-    if uri is not None:
-        # Make sure we can create a dataset first
+    archive_less_mature: Optional[Union[bool, Iterable[str]]]=None
+) -> int:
+    if uri is None:
+        raise IndexingException("Failed to get URI from metadata doc")
+    # Make sure we can create a dataset first
+    try:
+        ds, err = doc2ds(metadata, uri)
+    except ValueError as e:
+        raise IndexingException(
+            f"Exception thrown when trying to create dataset: '{e}'\n The URI was {uri}"
+        )
+    if ds is None:
+        raise IndexingException(
+            f"Failed to create dataset with error {err}\n The URI was {uri}"
+        )
+
+    if archive_less_mature:
+        if archive_less_mature == True:
+            # if set explicitly to True, default to [region_code, time]
+            archive_less_mature = ["region_code", "time"]
         try:
-            ds, err = doc2ds(metadata, uri)
-        except ValueError as e:
+            dupe_query = {k: getattr(ds.metadata, k) for k in archive_less_mature}
+        except AttributeError as e:
             raise IndexingException(
-                f"Exception thrown when trying to create dataset: '{e}'\n The URI was {uri}"
+                f"Cannot extract matching value from dataset for maturity check: {e}\n The URI was {uri}"
+            )
+    else:
+        dupe_query = {}
+
+    with dc.index.transaction():
+        # Process in a transaction
+        archive_ids = []
+        added = False
+        updated = False
+        if archive_less_mature:
+            dupes =  dc.index.datasets.search(
+                product=ds.type,
+                **dupe_query
             )
+            for dupe in dupes:
+                if dupe.id == ds.id:
+                    # Same dataset, for update.  Ignore
+                    continue
+                if dupe.metadata.dataset_maturity <= ds.metadata.dataset_maturity:
```

Comment:
I'm curious: How does this line work?
I see that typical values are 'final|interim|nrt' but are these a controlled vocabulary and how/where are their order defined?

Reply:
The "allowed" values are defined in the EO3 definition at [datacube/index/default-metadata-types.yaml:42](https://github.com/opendatacube/datacube-core/blob/develop/datacube/index/default-metadata-types.yaml#L41-L44) and are: `nrt, interim, final` listed in of increasing maturity and decreasing alphabetic order.

There is no code that enforces the use of only these values, so if somebody decided to use new dataset maturity values that didn't neatly sort in reverse alphabetic order, they would have problems if they tried to use this code.  But the supported values are well documented (and have been for quite some time), and the `dataset_maturity` metadata element is defined within the `dea` namespace, so I feel comfortable with this implementation.